File: portfolio-pnl-simulator/src/core/workers/calculator.py
import numpy as np
from scipy.stats import norm

from src.core.utils import get_forward_price
from src.core.models.future import Future
from src.core.models.market import Market
from src.core.models.option import Option
from src.core.models.optiontype import OptionType
from src.core.models.risk import Risk
from src.core.models.security import Security

def get_forward_strike(market: Market, strike: float, expiry: int) -> float:
    tte = expiry / 365
    return strike * np.exp(market.interest_rate * tte)

def price_future(future: Future, market: Market) -> float:
    tte = future.expiry / 365
    return get_forward_price(market.spot, market.interest_rate, future.expiry)

def get_option_pricer_inputs(option: Option, market: Market):
    tte = option.expiry / 365
    forward = market.spot * np.exp(market.interest_rate * tte)
    sigma = market.vol_surface.get_vol(option.strike, forward, option.expiry)
    d1 = (np.log(forward / option.strike) + 0.5 * sigma**2 * tte) / (sigma * np.sqrt(tte))
    d2 = d1 - sigma * np.sqrt(tte)
    df = np.exp(-market.interest_rate * tte)
    return tte, forward, sigma, d1, d2, df

def price_option(option: Option, market: Market) -> float:
    tte, forward, _, d1, d2, _ = get_option_pricer_inputs(option, market)
    if option.option_type == OptionType.CALL:
        price = np.exp(-market.interest_rate * tte) * (forward * norm.cdf(d1) - option.strike * norm.cdf(d2))
    else:
        price = np.exp(-market.interest_rate * tte) * (option.strike * norm.cdf(-d2) - forward * norm.cdf(-d1))
    return price

def calculate_risk(option: Option, market: Market) -> Risk:
    tte, forward, sigma, d1, _, df = get_option_pricer_inputs(option, market)
    delta = df * (norm.cdf(d1) if option.option_type == OptionType.CALL else -norm.cdf(-d1))
    gamma = df * norm.pdf(d1) / (forward * sigma * np.sqrt(tte))
    vega = df * forward * norm.pdf(d1) * np.sqrt(tte)
    theta_time = -df * forward * norm.pdf(d1) * sigma / (2.0 * np.sqrt(tte))
    if option.option_type == OptionType.CALL:
        theta_carry = -market.interest_rate * df * forward * norm.cdf(d1)
    else:
        theta_carry = +market.interest_rate * df * forward * norm.cdf(-d1)
    theta = theta_time + theta_carry
    return Risk(delta=delta, gamma=gamma, vega=vega, theta=theta)

def calculate_portfolio_risk(positions: list[Security], market: Market) -> Risk:
    total_delta = 0.0
    total_gamma = 0.0
    total_vega = 0.0
    total_theta = 0.0
    for position in positions:
        if isinstance(position, Option):
            risk = calculate_risk(position, market)
            total_delta += risk.delta * position.lot_size
            total_gamma += risk.gamma * position.lot_size
            total_vega += risk.vega * position.lot_size
            total_theta += risk.theta * position.lot_size
        elif isinstance(position, Future):
            total_delta += 1.0 * position.lot_size
    return Risk(delta=total_delta, gamma=total_gamma, vega=total_vega, theta=total_theta)

File: temp-options-market-making/src/core/workers/calculator.cpp
#include "core/workers/calculator.hpp"
#include "core/utils.hpp"
#include <cmath>

namespace omm::core::workers {

constexpr double PI = 3.14159265358979323846;
constexpr double SQRT_2PI = 2.50662827463100050241;

// Standard normal CDF
static double normCdf(double x) {
    return 0.5 * (1.0 + std::erf(x / std::sqrt(2.0)));
}

// Standard normal PDF
static double normPdf(double x) {
    return std::exp(-0.5 * x * x) / SQRT_2PI;
}

double Calculator::priceFuture(const omm::core::models::Future& future, const omm::core::models::Market& market) {
    double tte = std::stod(future.expiry) / 365.0;
    return Utils::getForwardPrice(market.spot, market.interestRate, std::stod(future.expiry));
}

OptionPricerInputs Calculator::getOptionPricerInputs(
    const omm::core::models::Option& option,
    const omm::core::models::Market& market
) {
    double tte = option.expiry / 365.0;
    double forward = market.spot * std::exp(market.interestRate * tte);
    double sigma = market.volSurface->getVol(option.strike, forward, option.expiry);
    
    double d1 = (std::log(forward / option.strike) + 0.5 * sigma * sigma * tte) / (sigma * std::sqrt(tte));
    double d2 = d1 - sigma * std::sqrt(tte);
    double df = std::exp(-market.interestRate * tte);
    
    return OptionPricerInputs{tte, forward, sigma, d1, d2, df};
}

double Calculator::priceOption(const omm::core::models::Option& option, const omm::core::models::Market& market) {
    OptionPricerInputs inputs = getOptionPricerInputs(option, market);
    
    if (option.optionType == omm::core::models::OptionType::CALL) {
        return inputs.df * (inputs.forward * normCdf(inputs.d1) - option.strike * normCdf(inputs.d2));
    } else {
        return inputs.df * (option.strike * normCdf(-inputs.d2) - inputs.forward * normCdf(-inputs.d1));
    }
}

omm::core::models::Risk Calculator::calculateRisk(
    const omm::core::models::Option& option,
    const omm::core::models::Market& market
) {
    OptionPricerInputs inputs = getOptionPricerInputs(option, market);
    
    double delta, theta;
    if (option.optionType == omm::core::models::OptionType::CALL) {
        delta = inputs.df * normCdf(inputs.d1);
    } else {
        delta = -inputs.df * normCdf(-inputs.d1);
    }
    
    double gamma = inputs.df * normPdf(inputs.d1) / (inputs.forward * inputs.sigma * std::sqrt(inputs.tte));
    double vega = inputs.df * inputs.forward * normPdf(inputs.d1) * std::sqrt(inputs.tte);
    
    double thetaTime = -inputs.df * inputs.forward * normPdf(inputs.d1) * inputs.sigma / (2.0 * std::sqrt(inputs.tte));
    double thetaCarry;
    if (option.optionType == omm::core::models::OptionType::CALL) {
        thetaCarry = -market.interestRate * inputs.df * inputs.forward * normCdf(inputs.d1);
    } else {
        thetaCarry = market.interestRate * inputs.df * inputs.forward * normCdf(-inputs.d1);
    }
    theta = thetaTime + thetaCarry;
    
    return omm::core::models::Risk(delta, gamma, vega, theta);
}

omm::core::models::Risk Calculator::calculatePortfolioRisk(
    const std::vector<std::shared_ptr<omm::core::models::Security>>& positions,
    const omm::core::models::Market& market
) {
    double totalDelta = 0.0, totalGamma = 0.0, totalVega = 0.0, totalTheta = 0.0;
    
    for (const auto& position : positions) {
        // Skip null positions
        if (!position) continue;
        
        // Try to cast as Option
        auto* optPtr = dynamic_cast<omm::core::models::Option*>(position.get());
        if (optPtr != nullptr) {
            omm::core::models::Risk risk = calculateRisk(*optPtr, market);
            totalDelta += risk.delta * optPtr->lotSize;
            totalGamma += risk.gamma * optPtr->lotSize;
            totalVega += risk.vega * optPtr->lotSize;
            totalTheta += risk.theta * optPtr->lotSize;
            continue;
        }
        
        // Try to cast as Future
        auto* futPtr = dynamic_cast<omm::core::models::Future*>(position.get());
        if (futPtr != nullptr) {
            totalDelta += 1.0 * futPtr->lotSize;
        }
    }
    
    return omm::core::models::Risk(totalDelta, totalGamma, totalVega, totalTheta);
}

double Calculator::getForwardStrike(const omm::core::models::Market& market, double strike, double expiry) {
    double tte = expiry / 365.0;
    return strike * std::exp(market.interestRate * tte);
}

}  // namespace omm::core::workers
